crates/driver/src/domain/competition/solution/settlement.rs

```diff
@@ -1,12 +1,314 @@
+            let trade::Execution { sell, buy } =
+                trade.execution(&solution).map_err(|_| Error::AssetFlow)?;
+            *flow.entry(sell.token).or_default() += util::conv::u256::to_big_int(sell.amount);
+            // Within the settlement contract, the orders which buy ETH are wrapped into
+            // WETH, and hence contribute to WETH flow.
+            *flow.entry(buy.token.wrap(solution.weth)).or_default() -=
+                util::conv::u256::to_big_int(buy.amount);
+        }
+
+        if flow.values().any(|v| v.is_negative()) {
+            return Err(Error::AssetFlow);
+        }
+
+        // Internalization rule: check that internalized interactions only use trusted
+        // tokens.
+        if !solution
+            .interactions
+            .iter()
+            .filter(|interaction| interaction.internalize())
+            .all(|interaction| {
+                interaction
+                    .inputs()
+                    .iter()
+                    .all(|asset| auction.is_trusted(asset.token))
+            })
+        {
+            return Err(Error::UntrustedInternalization);
+        }
+
+        // Encode the solution into a settlement.
+        let boundary = boundary::Settlement::encode(eth, &solution, auction).await?;
+        Self::new(
+            Id::random(),
+            [(solution.id, solution)].into(),
+            boundary,
+            eth,
+            simulator,
+        )
+        .await
+    }
+
+    /// Create a new settlement and ensure that it is valid.
+    async fn new(
+        id: Id,
+        solutions: HashMap<solution::Id, Solution>,
+        settlement: boundary::Settlement,
+        eth: &Ethereum,
+        simulator: &Simulator,
+    ) -> Result<Self, Error> {
+        let (access_list, gas) = Self::simulate(
+            id,
+            solutions.clone(),
+            settlement.clone(),
+            eth,
+            simulator,
+            Internalization::Enable,
+        )
+        .await?;
+
+        // Ensure that the solver has sufficient balance for the settlement to be mined.
+        if eth.balance(settlement.solver).await? < gas.required_balance() {
+            return Err(Error::InsufficientBalance);
+        }
+
+        // Some rules which are enforced by the settlement contract for non-internalized
+        // interactions are not enforced for internalized interactions (in order to save
+        // gas). However, publishing a settlement with interactions that violate
+        // these rules constitutes a punishable offense for the solver, even if
+        // the interactions are internalized. To ensure that this doesn't happen, check
+        // that the settlement simulates even when internalizations are disabled.
+        if solutions
+            .values()
+            .flat_map(|solution| solution.interactions.iter())
+            .any(|interaction| interaction.internalize())
+        {
+            Self::simulate(
```

Comment:
I think there might be a simple way to avoid a bunch of node requests and thus reducing latency if it's needed in the future by passing in an `Option<AccessList>` into `Self::simulate()`.
In the second simulation (with `Internalization::Disable`) you already know which access list is required to make the tx work for smart contract wallets so you could easily pass that access list into the second simulation.
This idea assumes that `eth_createAccessList` ignores useless access list entries attached to the tx (which I'm not 100% sure is the case).

Reply:
Like this? https://github.com/cowprotocol/services/pull/1490/commits/a0963a71f0c9180816fd7cf6518ef84b56d3a809

Reply:
Yes. This is even nicer since we don't have to depend on `eth_createAccessList` dropping unnecessary storage items for us. 👍